consensus/src/chained_bft/block_storage/block_store.rs

```diff
@@ -209,21 +218,30 @@ impl<T: Payload> BlockStore<T> {
                 return Err(e);
             }
         };
-        let compute_res = self
-            .state_computer
-            .compute(
-                parent_id,
-                block.id(),
-                block.payload().unwrap_or(&T::default()),
-            )
-            .await
-            .with_context(|e| format!("Execution failure for block {}: {:?}", block, e))?;
-
-        self.storage
-            .save_tree(vec![block.clone()], vec![])
-            .with_context(|e| format!("Insert block failed with {:?} when saving block", e))?;
-        let new_block = ExecutedBlock::new(block, compute_res);
-        self.inner.write().unwrap().insert_block(new_block)
+        // Reconfiguration rule - if a block is a child of reconfiguration, it needs to be empty
+        // So we roll over the executed state until it's committed and we start new epoch.
+        let parent_state = parent_block.compute_result();
+        let compute_res = if parent_state.executed_state.validators.is_some() {
+            ensure!(
+                block.payload().filter(|p| **p != T::default()).is_none(),
```

Comment:
Why is it `**p`?

Reply:
well, payload() returns a Option<&T> and filter takes a P: FnOnce(&T) for Option<T> so p becomes &&T